cmd/nerdctl/ipfs_push.go

```diff
@@ -0,0 +1,144 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/images/converter"
+	"github.com/containerd/nerdctl/pkg/ipfs"
+	"github.com/containerd/stargz-snapshotter/estargz"
+	"github.com/containerd/stargz-snapshotter/estargz/zstdchunked"
+	estargzconvert "github.com/containerd/stargz-snapshotter/nativeconverter/estargz"
+	httpapi "github.com/ipfs/go-ipfs-http-client"
+	digest "github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/sirupsen/logrus"
+
+	"github.com/spf13/cobra"
+)
+
+func newIPFSPushCommand() *cobra.Command {
+	var ipfsPushCommand = &cobra.Command{
+		Use:           "push CID",
+		Short:         "push an image to IPFS",
+		RunE:          ipfsPushAction,
+		SilenceUsage:  true,
+		SilenceErrors: true,
+	}
+
+	// #region platform flags
+	// platform is defined as StringSlice, not StringArray, to allow specifying "--platform=amd64,arm64"
+	ipfsPushCommand.PersistentFlags().StringSlice("platform", nil, "Push content for a specific platform")
+	ipfsPushCommand.RegisterFlagCompletionFunc("platform", shellCompletePlatforms)
+	ipfsPushCommand.PersistentFlags().Bool("all-platforms", false, "Push content for all platforms")
+	// #endregion
+
+	ipfsPushCommand.PersistentFlags().Bool("estargz", true, "Convert the image into eStargz")
```

Comment:
If we want `--estargz` flag here, we should add that to the regular `nerdctl image push` too, and deduplicate the code across `nerdctl image push` and `nerdctl ipfs push`. 
(But do we really need `nerdctl ipfs *` commands?)